File: cockfuck.cpp
/*
 *   DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE 
 *            Version 2, December 2004 
 *
 * Copyright (C) 2004 Sam Hocevar <sam@hocevar.net> 
 *
 * Everyone is permitted to copy and distribute verbatim or modified 
 * copies of this license document, and changing it is allowed as long 
 * as the name is changed. 
 *
 *            DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE 
 *   TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION 
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 *
 *
 * cockfuck.rs - cockfuck interpreter in rust
 */

#include "cockfuck.hpp"

void add_cmd(ParseState &p, ProgStep &&t, bool force) {
    if (! force && (t.n == 0)) {
        parse_err(p);
    } else {
        p.cmds.push_back(t);
        parse_set_at(p);
    }
}

void parse_err(ParseState &p) {
    if (fseek(p.infile, p.at + 1, SEEK_SET) < 0) {
        perror("Seek error");
        exit(-1);
    }
    parse_set_at(p);
}

void parse_set_at(ParseState &p) {
    p.at = ftell(p.infile);
}

void parse_back_up(ParseState &p) {
    if (fseek(p.infile, -1, SEEK_CUR) < 0) {
        perror("Seek error");
        exit(-1);
    }
    parse_set_at(p);
}

void parse_infile(ParseState &p) {
    int c;
    while ( (c = fgetc(p.infile)) != EOF ) {
        switch ((char) c) {
            case '8': parse_hanging(p, 0); break;
            case 'B': parse_cupped(p, 0); break;
            case '-': parse_sound(p, 1); break;
            case '~': parse_jizz(p, 1); break;
            case '`': parse_cglans(p); break;
            default : parse_err(p);
        }
    }
    if (ferror(p.infile)) {
        perror("Error reading program file");
        exit(-1);
    }
}

void parse_hanging(ParseState &p, unsigned n) {
    int c = fgetc(p.infile);
    if (c == EOF) {
        return;
    }
    switch ((char) c) {
        case '=': parse_hanging(p, n+1); break;
        case '>': add_cmd(p, { ptr_inc, n }); break;
        case 'D': add_cmd(p, { ptr_dec, n }); break;
        case ',': p.lcStack.push_back(p.cmds.size()); add_cmd(p, { jump_fwd, n }, true); break;
        case '`':
        case '~':
        case '-':
        case '8':
        case 'B': parse_err(p); break;
        default : parse_hanging(p, n);
    }
}

void parse_cupped(ParseState &p, unsigned n) {
    int c = fgetc(p.infile);
    if (c == EOF) {
        return;
    }
    switch ((char) c) {
        case '=': parse_cupped(p, n+1); break;
        case '>': add_cmd(p, { data_inc, n }); break;
        case 'D': add_cmd(p, { data_dec, n }); break;
        case ',':
        case '`':
        case '~':
        case '-':
        case '8':
        case 'B': parse_err(p); break;
        default : parse_cupped(p, n);
    }
}

void parse_sound(ParseState &p, unsigned n) {
    int c = fgetc(p.infile);
    if (c == EOF) {
        add_cmd(p, { take_in, n });
        return;
    }
    switch ((char) c) {
        case '-': parse_sound(p, n+1); break;
        case '8':
        case 'B':
        case 'D':
        case '>':
        case '=':
        case ',':
        case '`':
        case '~': add_cmd(p, { take_in, n }); parse_back_up(p); break;
        default : parse_sound(p, n);
    }
}

void parse_jizz(ParseState &p, unsigned n) {
    int c = fgetc(p.infile);
    if (c == EOF) {
        add_cmd(p, { put_out, n });
        return;
    }
    switch ((char) c) {
        case '~': parse_jizz(p, n+1); break;
        case '8':
        case 'B':
        case 'D':
        case '>':
        case '=':
        case ',':
        case '`':
        case '-': add_cmd(p, { put_out, n }); parse_back_up(p); break;
        default : parse_jizz(p, n);
    }
}

void parse_cglans(ParseState &p) {
    int c = fgetc(p.infile);
    if (c == EOF) {
        return;
    }
    switch ((char) c) {
        case '=': parse_cglans(p); break;
        case 'D':
        case ',':
        case '`':
        case '~':
        case '-':
        case '8':
        case 'B': parse_err(p); break;
        case '>': if (p.lcStack.size() < 1) {
                      printf("unbalanced chopticock! program is invalid.\n");
                      exit(-1);
                  } else {
                      unsigned last = p.lcStack.back();
                      p.lcStack.pop_back();
                      unsigned ppos = p.cmds.size();
                      p.cmds[last].n = ppos;
                      add_cmd(p, { jump_back, last }, true);
                  }
                  break;
        default : parse_cglans(p);
    }
}

int main(int argc, char **argv) {
    if (argc != 2) {
        printf("Usage: %s <infile>\n",argv[0]);
        exit(-1);
    }

    ParseState pState;
    pState.at = 0;
    if ( (pState.infile = fopen(argv[1],"r")) == NULL ) {
        perror("Could not open input file");
        exit(-1);
    }

    parse_infile(pState);

    if (pState.lcStack.size() > 0) {
        printf("unbalanced chopticock! program is invalid.\n");
        exit(-1);
    }

    pState.cmds.push_back({ halt, 0 });
    std::vector<ProgStep> cmds = pState.cmds;

    bool cont = true;
    int c = 0;
    ProgStep s;

    while (cont) {
        s = cmds[cPtr];
        switch(s.op) {
            case data_inc : array[dPtr] += s.n;
                            break;
            case data_dec : array[dPtr] -= s.n;
                            break;
            case ptr_inc  : dPtr += s.n;
                            break;
            case ptr_dec  : dPtr -= s.n;
                            break;
            case take_in  : for (unsigned i=0; i<s.n; i++) {
                                c = getchar();
                            }
                            array[dPtr] = (c == EOF) ? 0 : c;
                            break;
            case put_out  : for (unsigned i=0; i<s.n; i++) {
                                putc(array[dPtr], stdout);
                            }
                            break;
            case jump_fwd : if (array[dPtr] == 0) {
                                cPtr = s.n;
                            }
                            break;
            case jump_back: if (array[dPtr] != 0) {
                                cPtr = s.n;
                            }
                            break;
            case halt     : cont = false;
        }
        cPtr++;
    }

    return 0;
}

File: cockfuck.rs
/*
 *   DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE 
 *            Version 2, December 2004 
 *
 * Copyright (C) 2004 Sam Hocevar <sam@hocevar.net> 
 *
 * Everyone is permitted to copy and distribute verbatim or modified 
 * copies of this license document, and changing it is allowed as long 
 * as the name is changed. 
 *
 *            DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE 
 *   TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION 
 *
 *  0. You just DO WHAT THE FUCK YOU WANT TO.
 *
 *
 * cockfuck.rs - cockfuck interpreter in rust
 */

use std::os;
use std::io::{stdio, File, SeekSet, SeekCur};

#[deriving(Clone, Show)]
enum StepType {
    DataInc(u8),
    DataDec(u8),
    PtrInc(u16),
    PtrDec(u16),
    TakeIn(uint),
    PutOut(uint),
    JumpFwd(uint),
    JumpBack(uint),
    Halt
}

// state for use while parsing
struct ParseInfo {
    f : File,
    at : i64,
    lcStack : Vec<uint>,
    cmds : Vec<StepType>
}

fn parse_err(p : &mut ParseInfo) {
    match p.f.seek(p.at + 1, SeekSet) {
        Err(e) => fail!("Seek failed: {}",e),
        _ => ()
    };
    parse_set_at(p);
}

fn parse_set_at(p : &mut ParseInfo) {
    p.at = match p.f.tell() {
        Err(e) => fail!("Tell failed: {}",e),
        Ok(t) => t as i64
    }
}

fn parse_back_up(p : &mut ParseInfo) {
    match p.f.seek(-1, SeekCur) {
        Err(e) => fail!("Seek failed: {}",e),
        _ => ()
    };
    parse_set_at(p);
}


fn parse_infile(p : &mut ParseInfo) {
    loop {
        match p.f.read_byte() {
            Err(_) => break,
            Ok(b) => match b as char {
                '8' => parse_hanging(p,0),
                'B' => parse_cupped(p,0),
                '-' => parse_sound(p,1),
                '~' => parse_jizz(p,1),
                '`' => parse_cglans(p),
                _   => parse_err(p)
            }
        }
    }
}

// 8
// =
// >
// D
// B
// -
// ~
// `
// ,

fn add_cmd(p : &mut ParseInfo, t : StepType, n : uint) {
    if n == 0 {
        parse_err(p);
    } else {
        p.cmds.push(t);
        parse_set_at(p);
    }
}

fn parse_hanging(p : &mut ParseInfo, n : uint) {
    match p.f.read_byte() {
        Err(_) => return,
        Ok(b) => match b as char {
            '=' => parse_hanging(p, n+1),
            '>' => add_cmd(p, PtrInc(n as u16), n),
            'D' => add_cmd(p, PtrDec(n as u16), n),
            ',' => { p.lcStack.push(p.cmds.len()); add_cmd(p, JumpFwd(n), 1); },
            '`' | '~' | '-' | '8' | 'B' => parse_err(p),
            _   => parse_hanging(p,n)
        }
    }
}

fn parse_cupped(p : &mut ParseInfo, n : uint) {
    match p.f.read_byte() {
        Err(_) => return,
        Ok(b) => match b as char {
            '=' => parse_cupped(p, n+1),
            '>' => add_cmd(p, DataInc(n as u8), n),
            'D' => add_cmd(p, DataDec(n as u8), n),
            ',' | '`' | '~' | '-' | '8' | 'B' => parse_err(p),
            _   => parse_cupped(p,n)
        }
    }
}

fn parse_sound(p : &mut ParseInfo, n : uint) {
    match p.f.read_byte() {
        Err(_) => return,
        Ok(b) => match b as char {
            '-' => parse_sound(p, n+1),
            _   => { p.cmds.push(TakeIn(n)); parse_back_up(p); }
        }
    }
}

fn parse_jizz(p : &mut ParseInfo, n : uint) {
    match p.f.read_byte() {
        Err(_) => return,
        Ok(b) => match b as char {
            '~' => parse_jizz(p, n+1),
            _   => { p.cmds.push(PutOut(n)); parse_back_up(p); }
        }
    }
}

fn parse_cglans(p : &mut ParseInfo) {
    match p.f.read_byte() {
        Err(_) => return,
        Ok(b) => match b as char {
            '=' => parse_cglans(p),
            'D' | ',' | '`' | '~' | '-' | '8' | 'B' => parse_err(p),
            '>' => {
                let lstck = match p.lcStack.pop() {
                    None => fail!("unbalanced chopticock! program is invalid"),
                    Some(t) => t
                };
                let ppos = p.cmds.len();
                *(p.cmds.get_mut(lstck)) = JumpFwd(ppos);
                add_cmd(p, JumpBack(lstck), 1);
            }
            _   => parse_cglans(p)
        }
    }
}

fn main() {
    let args : Vec<String> = os::args();
    let program : &String = args.get(0);
    if args.len() != 2 {
        println!("Usage: {} <infile>",program);
        os::set_exit_status(-1);
        return;
    }

    let inf = &Path::new(args.get(1).clone());
    let infile = match File::open(inf) {
        Ok(f) => f,
        Err(e) => fail!("Could not open input file {}: {}",args.get(1),e)
    };

    let mut pInfo = ParseInfo { f : infile, at : 0, lcStack : vec!(), cmds : vec!() };

    parse_infile(&mut pInfo);

    if pInfo.lcStack.len() > 0 {
        fail!("unbalanced chopticock! program is invalid");
    }

    pInfo.cmds.push(Halt);
    let cmds = pInfo.cmds.as_slice();

    let mut dPtr : u16 = 0;
    let mut cPtr : uint = 0;
    let mut array : [u8, ..65536] = [0, ..65536];
    let mut sIn = stdio::stdin_raw();
    let mut sOut = stdio::stdout();

    loop {
        match &cmds[cPtr] {
            &Halt => break,
            &DataInc(d) => array[dPtr as uint] += d,
            &DataDec(d) => array[dPtr as uint] -= d,
            &PtrInc(p)  => dPtr += p,
            &PtrDec(p)  => dPtr -= p,
            &TakeIn(n)  => array[dPtr as uint] =
                            match sIn.read_exact(n) {
                                Ok(v) => *v.get(n-1) as u8,
                                Err(_)=> 0
                            },
            &PutOut(n)  => sOut.write_str( String::from_char(n, array[dPtr as uint] as char).as_slice() ).unwrap_or(()),
            &JumpFwd(n) => if array[dPtr as uint] == 0 { cPtr = n },
            &JumpBack(n)=> if array[dPtr as uint] != 0 { cPtr = n }
        };

        cPtr += 1;
    }
}
